pkg/embeddedcluster/upgrade.go

```diff
@@ -0,0 +1,380 @@
+package embeddedcluster
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/mholt/archiver/v3"
+	embeddedclusterv1beta1 "github.com/replicatedhq/embedded-cluster-kinds/apis/v1beta1"
+	embeddedclustertypes "github.com/replicatedhq/embedded-cluster-kinds/types"
+	dockerregistrytypes "github.com/replicatedhq/kots/pkg/docker/registry/types"
+	"github.com/replicatedhq/kots/pkg/imageutil"
+	"github.com/replicatedhq/kots/pkg/k8sutil"
+	registrytypes "github.com/replicatedhq/kots/pkg/registry/types"
+	"github.com/replicatedhq/kots/pkg/util"
+	kotsv1beta1 "github.com/replicatedhq/kotskinds/apis/kots/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+const (
+	clusterUpgradeTimeout = 15 * time.Minute
+)
+
+// startClusterUpgrade will create a new installation with the provided config.
+func startClusterUpgrade(
+	ctx context.Context, newcfg embeddedclusterv1beta1.ConfigSpec,
+	artifacts *embeddedclusterv1beta1.ArtifactsLocation,
+	registrySettings registrytypes.RegistrySettings,
+	license kotsv1beta1.License, versionLabel string,
+) error {
+	// TODO(upgrade): put a lock here to prevent multiple upgrades at the same time
+
+	kbClient, err := k8sutil.GetKubeClient(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get kubeclient: %w", err)
+	}
+
+	k8sClient, err := k8sutil.GetClientset()
+	if err != nil {
+		return fmt.Errorf("failed to get clientset: %w", err)
+	}
+
+	current, err := GetCurrentInstallation(ctx, kbClient)
+	if err != nil {
+		return fmt.Errorf("failed to get current installation: %w", err)
+	}
+	newins := &embeddedclusterv1beta1.Installation{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: embeddedclusterv1beta1.GroupVersion.String(),
+			Kind:       "Installation",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: time.Now().Format("20060102150405"),
+			Labels: map[string]string{
+				"replicated.com/disaster-recovery": "ec-install",
+			},
+		},
+		Spec: embeddedclusterv1beta1.InstallationSpec{
+			ClusterID:                 current.Spec.ClusterID,
+			MetricsBaseURL:            current.Spec.MetricsBaseURL,
+			HighAvailability:          current.Spec.HighAvailability,
+			AirGap:                    current.Spec.AirGap,
+			Artifacts:                 artifacts,
+			Config:                    &newcfg,
+			EndUserK0sConfigOverrides: current.Spec.EndUserK0sConfigOverrides,
+			BinaryName:                current.Spec.BinaryName,
+			LicenseInfo:               &embeddedclusterv1beta1.LicenseInfo{IsDisasterRecoverySupported: license.Spec.IsDisasterRecoverySupported},
+		},
```

Comment:
At some point we should start doing spec = current.spec and then just replacing the artifacts and configs